evm/src/invariant_fuzz.rs

```diff
@@ -0,0 +1,363 @@
+        let mid = self.cases.len() / 2;
+        self.cases.get(mid).map(|c| c.gas).unwrap_or_default()
+    }
+
+    /// Returns the average gas use of all test cases
+    pub fn mean_gas(&self) -> u64 {
+        if self.cases.is_empty() {
+            return 0
+        }
+
+        (self.cases.iter().map(|c| c.gas as u128).sum::<u128>() / self.cases.len() as u128) as u64
+    }
+
+    pub fn highest(&self) -> Option<&FuzzCase> {
+        self.cases.last()
+    }
+
+    pub fn lowest(&self) -> Option<&FuzzCase> {
+        self.cases.first()
+    }
+
+    pub fn highest_gas(&self) -> u64 {
+        self.highest().map(|c| c.gas).unwrap_or_default()
+    }
+
+    pub fn lowest_gas(&self) -> u64 {
+        self.lowest().map(|c| c.gas).unwrap_or_default()
+    }
+}
+
+/// Data of a single fuzz test case
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct InvariantFuzzCase {
+    /// The calldata used for this fuzz test
+    pub calldata: Bytes,
+    // Consumed gas
+    pub gas: u64,
+}
+
+pub fn invariant_strat(
+    depth: usize,
+    contracts: BTreeMap<Address, (String, Abi)>,
+) -> BoxedStrategy<Vec<(Address, Bytes)>> {
+    let iters = 1..depth + 1;
+    proptest::collection::vec(gen_call(contracts), iters).boxed()
+}
+
+fn gen_call(contracts: BTreeMap<Address, (String, Abi)>) -> BoxedStrategy<(Address, Bytes)> {
+    let random_contract = select_random_contract(contracts);
+    random_contract
+        .prop_flat_map(move |(contract, abi)| {
+            let func = select_random_function(abi);
+            func.prop_flat_map(move |func| fuzz_calldata(contract, func))
+        })
+        .boxed()
+}
+
+fn select_random_contract(
+    contracts: BTreeMap<Address, (String, Abi)>,
+) -> impl Strategy<Value = (Address, Abi)> {
+    let selectors = any::<prop::sample::Selector>();
+    selectors.prop_map(move |selector| {
+        let res = selector.select(&contracts);
+        (*res.0, res.1 .1.clone())
+    })
+}
+
+fn select_random_function(abi: Abi) -> impl Strategy<Value = Function> {
+    let selectors = any::<prop::sample::Selector>();
+    let possible_funcs: Vec<ethers::abi::Function> = abi
+        .functions()
+        .filter(|func| {
+            !matches!(
+                func.state_mutability,
+                ethers::abi::StateMutability::Pure | ethers::abi::StateMutability::View
+            )
+        })
+        .cloned()
+        .collect();
+    selectors.prop_map(move |selector| {
+        let func = selector.select(&possible_funcs);
+        func.clone()
+    })
+}
+
+/// Given a function, it returns a proptest strategy which generates valid abi-encoded calldata
+/// for that function's input types.
+pub fn fuzz_calldata(addr: Address, func: Function) -> impl Strategy<Value = (Address, Bytes)> {
+    // We need to compose all the strategies generated for each parameter in all
+    // possible combinations
+    let strats = func.inputs.iter().map(|input| fuzz_param(&input.kind)).collect::<Vec<_>>();
```

Comment:
@brockelmore Have we talked about additional invariant-specific strategies for generating calldata? I think it'd be interesting to add:
1. All stack/memory values, and/or
2. The output of calling view methods which have no inputs

to the dict for the lifetime of one invariant run, then remove them for the subsequent run to avoid flooding the dict. Maybe a separate issue/PR though.